pkg/proxydetect/proxydetect_windows.go

```diff
@@ -0,0 +1,228 @@
+	} else {
+		defer func() {
+			if err := ieProxy.Free(); err != nil {
+				log.Errorf("failed to free resources after winapi call: %s", err.Error())
+			}
+		}()
+
+		if ieProxy.FAutoDetect {
+			proxyInfo, err := getProxyInfoForUrl(u,
+				&winapi.HttpAutoProxyOptions{
+					DwFlags:                winapi.WINHTTP_AUTOPROXY_AUTO_DETECT,
+					DwAutoDetectFlags:      winapi.WINHTTP_AUTO_DETECT_TYPE_DHCP | winapi.WINHTTP_AUTO_DETECT_TYPE_DNS_A,
+					FAutoLogonIfChallenged: true,
+				})
+			if err == nil {
+				defer func() {
+					if err := proxyInfo.Free(); err != nil {
+						log.Errorf("WINHTTP_AUTOPROXY_AUTO_DETECT failed to free resources: %s", err.Error())
+					}
+				}()
+
+				proxy, err = matchProxyForURLAndBypassList(
+					u,
+					strings.Split(proxyInfo.LpszProxy.String(), ";"),
+					strings.Split(proxyInfo.LpszProxyBypass.String(), ";"),
+				)
+				if err == nil {
+					return proxy, nil
+				}
+			}
+			if err != ErrNotFound {
+				log.Errorf("WINHTTP_AUTOPROXY_AUTO_DETECT got error: %s", err.Error())
+			}
+		}
+
+		if ieProxy.LpszAutoConfigUrl.String() != "" {
+			proxyInfo, err := getProxyInfoForUrl(u,
+				&winapi.HttpAutoProxyOptions{
+					DwFlags:                winapi.WINHTTP_AUTOPROXY_CONFIG_URL,
+					LpszAutoConfigUrl:      ieProxy.LpszAutoConfigUrl,
+					FAutoLogonIfChallenged: true,
+				})
+			if err == nil {
+				defer func() {
+					if err := proxyInfo.Free(); err != nil {
+						log.Errorf("WINHTTP_AUTOPROXY_CONFIG_URL failed to free resources: %s", err.Error())
+					}
+				}()
+				proxy, err = matchProxyForURLAndBypassList(
+					u,
+					strings.Split(proxyInfo.LpszProxy.String(), ";"),
+					strings.Split(proxyInfo.LpszProxyBypass.String(), ";"),
+				)
+				if err == nil {
+					return proxy, nil
+				}
+			}
+
+			if err != ErrNotFound {
+				log.Errorf("WINHTTP_AUTOPROXY_CONFIG_URL got error: %s", err.Error())
+			}
+		}
+
+		proxy, err = matchProxyForURLAndBypassList(
+			u,
+			strings.Split(ieProxy.LpszProxy.String(), ";"),
+			strings.Split(ieProxy.LpszProxyBypass.String(), ";"),
+		)
+		if err == nil {
+			return proxy, nil
+		} else if err != ErrNotFound {
+			log.Errorf("WINHTTP_CURRENT_USER_IE_PROXY_CONFIG named proxy got error: %s", err.Error())
+		}
+	}
+
+	proxyInfo, err := winapi.HttpGetDefaultProxyConfiguration()
+	if err == nil {
+		defer func() {
+			if err := proxyInfo.Free(); err != nil {
+				log.Errorf("HttpGetDefaultProxyConfiguration failed to free resources: %s", err.Error())
+			}
+		}()
+
+		return matchProxyForURLAndBypassList(
+			u,
+			strings.Split(proxyInfo.LpszProxy.String(), ";"),
+			strings.Split(proxyInfo.LpszProxyBypass.String(), ";"),
+		)
+	} else if err != ErrNotFound {
+		log.Errorf("WinHttpGetDefaultProxyConfiguration got error: %s", err.Error())
+	}
+
+	return nil, ErrNotFound
+}
+
+func getProxyInfoForUrl(targetUrl *url.URL, autoProxyOptions *winapi.HttpAutoProxyOptions) (*winapi.HttpProxyInfo, error) {
+	userAgent, _ := windows.UTF16PtrFromString("cagent")
```

Comment:
Can we pass user agent string from Cagent.userAgent() method call? It contains more information

Reply:
done